rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,212 @@
+
+This RFC is part of a series to discuss a standardized way of specifying framing
+and codecs across sources and sinks. _Framing_ is concerned with turning a
+stream of bytes into byte frames (chunks of data that have finite size and
+contain complete messages), while a _Codec_ converts between a byte frame and
+structured data.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a component + transform into a component with codec config in places where the
+transform was merely used to convert between encoding formats, e.g. from `bytes`
+to `json` or vice versa.
+
+## Scope
+
+For this part, we focus on the source/decoding side and leave the sink/encoding
+side to a separate RFC.
+
+The scope of this RFC concerns at which level framing and decoding operates, how
+these framers and decoders can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for a framer or a
+decoder. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and decoding, such that each source may implement these in their own
+way. This has led to inconsistencies in available options and poses additional
+maintenance burden. Components have their own defaults, making
+behavior unpredictable and surprising users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+To expose framing and decoding to sources in a uniform way, we specify a common
+configuration struct:
+
+```rust
+pub struct DecodingConfig {
+    framing: Option<FramingConfig>,
+    decoding: Option<ParserConfig>,
+}
+```
+
+while the framing method and parser used to deserialize into a structured event
+can be selected from an enum:
+
+```rust
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum FramingConfig {
+    NewlineDelimited,
+    CharacterDelimited {
+        delimiter: char,
+        max_length: Option<usize>,
+    },
+    OctetCounting {
+        max_length: Option<usize>,
+    },
+    ...
+}
+
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ParserConfig {
+    Bytes,
+    Json {
+        ...
+    },
+    Syslog,
+    ...
+}
+```
+
+The `DecodingConfig` exposes a `build` method to create a Vector `Decoder` which
+implements the `tokio_util::codec::Decoder<Item = (Event, usize)>` trait. That
+way we can produce `Event`s from `ByteMut` by repeated calls to
+`decode`/`decode_eof` either on a byte stream or a byte message. The additional
+`usize` item conveys how many bytes were read to produce the particular `Event`,
+so that this information can be passed along to e.g. the internal event log.
+
+Internally, the `Decoder` holds a framer and a parser:
+
+```rust
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes> + Send + Sync>,
+    parser: Box<dyn Parser + Send + Sync>,
```

Comment:
Oh, that's really interesting! Haven't thought of it, indeed we'd know the exhaustive list of parsers/framers.

Looks like if we use an enum there, trivial implementations get inlined and the rest get `call`ed: [https://godbolt.org/...](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlJvoCqAZ0wAFAB4gA5AAYppAFZdSrZrVDEArkIKkR7ZATx1KmWugDCqVpoC2tZWfQAZPLUwA5OwCNMxEACsnKQADqhChMa0Vjb2ymERRnSu7l62vv5BepgGSbQMBMzEBDF2DsH6mIZRBUUEKZ4%2BfoEVhcWlca11DWkZLQCUeqiaxMgcUiGa3gDUOlqG09gAbmYEEAXEbsD9AKQAzABCOzIAgscnkzMExMyE06JFIsTTOwDsR6fTX9NUtNMhj0wEB2ACYAGz6KikabeACeBEwQhAL3BOwCB00AA40QARfrTAC0%2B2w0wASojNKwCPsLMtVtCILs9th9h8zq8cedzpdZtdNAsAGKoVAPYhPF7vN6c07nPC2EKse6A540Z5CkXKiVs74/P4AsVA0EQnJQmHwxHIo1ojHYgJ4wnEskUqk0un0BlMklvbU674AeQA1hA3WsNlsQCAqMRULYAPqaAhUTGx1jhISwiBwhFCfoAOjcREZ/V2n2%2BUq5HK5px5c35BGmByKovF3vLMtOcoVSoNKpIDabmtbpa%2Bv3%2BgOB4Mh0KzFpRYOtWNx%2BKJzKdQkp1L2tJW7umjMdQ5Ovu%2B7HrxEwKwNLz2OLN2dzRFjqAA7u50Iz8/RULHCH4P%2Belg/NBWFyCMaQANSqGlY2JYl90Oc5jy%2BQNgx3UNrnDSNozjBMkxTNMMyNAC/BEPMC1QIsSyPMtKxlWizlOWsFmbPwDlQcQtUQ759SeZE2PEGl0FhP4WOIWC6OlBiTk7RVRJ%2BPtRP4zjh11McDQnY1WFNGckTnBdbXtFcSXJdcXS3EMPQPSUVJ1SFcx4vx7PHHSqJ1Nt2Uk84zDsbsniEZTqK%2BdUIHVUTBi4r5G2ICAorC0gK08xi%2BWY5VsFoHzDx1Bz/F8kiEqraT5VkzVVVy4g0oy6zAtUhyNKnO9ZytdFFztZdHRMjdXTQyzV0ypDbGYAhkAQOc7OygKkN9USkRAYLsvxfZOVXbKnPUlz4pspDpojKKIHm68lpJFbavWiLfXctz6PcqRBlYaQAnkBxZHkVBpFJbR6yEYZRkwFE9k4eQCGkORi1IAMQDBEFcwCMFXgCABOTE9m4TEAkxV54fhlRpG4R7gZe6R5CRGRSCB57BjgWAkDQeU8HYMgKAgGmQjp5piE4AI9hJqg6YRMVKG8fHSG8NwilhaQAdIGnbFWP1aFYcXntILABvUdghfwc9qhWJElcwcQqgTcZJYLHIhdYPBvBuYhYSsLAhYw2wJZuvhVGAIRwLwTBnz9EIzGdvg6EYFh1Z4QPBBECQhaUYJVHUEAtB0FRLaRSBBlQEI8l1gk/SJnIqjyUxzA6Bw9jBUgnB6Jp/G4GRy4SSI6BLkAy9CcJG9oKv0maWvy8qao6FqdprDKFu%2B/zgf8jaeo3Eabua7rvRp%2Bb1udG6Wfeh7uvBi%2BkYxi4W77rxpXXqkROhumThob2XMZD3XBCD7UF/uhKxafpv7OHxd6dEB/HQfBnsV4uZuB7BBLXTgYIgjcE4JwTG2MpC41IE9OQpBT5ExACTMmIND5SBBMfVB6DSb/0GJeCIJhuBAA%3D).

I think in our case this would boil down to either
- a `call` to the location stored in the `Box` or
- first a matching on the enum tag and then `call`ing known locations

Performance wise, I'd expect no benefit for the one over the other.

---

On the other hand, I've been considering moving away from a fixed enum for the config (by using the [`typetag`](https://docs.rs/typetag/latest/typetag/) crate) since that would allow us to define the config options close to the implementation rather than having to group the configs in one place.

Therefore, I'd lean for using a trait object in both places.

What do you think?

Reply:
> Performance wise, I'd expect no benefit for the one over the other.

Historically, branch prediction has actually performed better than indirect jump prediction, though I don't know if that's still an issue with modern CPUs, and that is likely all lost in the noise given the other costs of framing and decoding.

> On the other hand, I've been considering moving away from a fixed enum for the config (by using the [`typetag`](https://docs.rs/typetag/latest/typetag/) crate) since that would allow us to define the config options close to the implementation rather than having to group the configs in one place.
> 
> Therefore, I'd lean for using a trait object in both places.

That would definitely point to using a trait object in both, yes. I was merely commenting on the asymmetry between the two.

Reply:
Makes sense, thanks for the perspective! Will adjust to be consistent

Reply:
Adjusted in 2d7c2b1ac1897be74e413fae2e9ba70e07965b71